src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,356 @@
+ * that uses this KV store to track consumption offsets for transforms.
+ */
+
+template<class T>
+concept SerdeSerializable = requires(T t, iobuf buf, iobuf_parser parser) {
+    serde::write_async(buf, t);
+    serde::read_async<T>(parser);
+    serde::from_iobuf<T>(std::move(buf));
+    serde::to_iobuf(t);
+};
+
+template<SerdeSerializable Key, SerdeSerializable Value>
+class distributed_kv_stm final : public persisted_stm<> {
+public:
+    explicit distributed_kv_stm(
+      size_t max_partitions, ss::logger& logger, raft::consensus* raft)
+      : persisted_stm<>("distributed_kv_stm.snapshot", logger, raft)
+      , _default_max_partitions(max_partitions)
+      , _is_routing_partition(_raft->ntp().tp.partition == routing_partition){};
+
+    ss::future<> start() override { co_await persisted_stm<>::start(); }
+    ss::future<> stop() override { co_await _gate.close(); }
+
+    ss::future<> apply(const model::record_batch& record_batch) override {
+        if (record_batch.header().type != model::record_batch_type::raft_data) {
+            co_return;
+        }
+        auto holder = _gate.hold();
+        auto units = co_await _snapshot_lock.hold_read_lock();
+        co_await record_batch.for_each_record_async(
+          [this](model::record r) -> ss::future<> {
+              auto key = reflection::from_iobuf<record_key>(r.release_key());
+              auto val = reflection::from_iobuf<record_value_wrapper>(
+                r.release_value());
+              if (key.type == record_type::repartitioning) {
+                  do_apply_repartition(std::move(val.actual_value));
+              } else if (key.type == record_type::coordinator_assignment) {
+                  do_apply_assignment(
+                    std::move(key.key_data), std::move(val.actual_value));
+              } else if (key.type == record_type::kv_data) {
+                  do_apply_kvs(
+                    std::move(key.key_data), std::move(val.actual_value));
+              } else {
+                  vlog(
+                    clusterlog.error,
+                    "skipped applying unknown record type: {}",
+                    static_cast<int8_t>(key.type));
+              }
+              return ss::now();
+          });
+    }
+
+    ss::future<>
+    apply_local_snapshot(stm_snapshot_header header, iobuf&& bytes) override {
+        auto holder = _gate.hold();
+        auto units = _snapshot_lock.hold_write_lock();
+
+        auto snapshot_offset = header.offset;
+        iobuf_parser parser(std::move(bytes));
+        auto snap = co_await serde::read_async<snapshot>(parser);
+
+        if (_is_routing_partition) {
+            _num_partitions = snap.num_partitions;
+            _coordinators.clear();
+            if (snap.coordinators) {
+                _coordinators = std::move(snap.coordinators.value());
+            }
+        }
+        _kvs = std::move(snap.kv_data);
+        set_next(model::next_offset(snapshot_offset));
```

Comment:
`set_next` is not required as it is already done in `persisted_stm`

Reply:
removed.